influxdb3_wal/src/object_store.rs

```diff
@@ -0,0 +1,907 @@
+use crate::serialize::verify_file_type_and_deserialize;
+use crate::snapshot_tracker::{SnapshotInfo, SnapshotTracker, WalPeriod};
+use crate::{
+    CatalogBatch, SnapshotDetails, Wal, WalConfig, WalContents, WalFileNotifier,
+    WalFileSequenceNumber, WalOp, WriteBatch,
+};
+use bytes::Bytes;
+use data_types::Timestamp;
+use futures_util::stream::StreamExt;
+use hashbrown::HashMap;
+use object_store::path::Path;
+use object_store::{ObjectStore, PutPayload};
+use observability_deps::tracing::error;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::sync::Mutex;
+use tokio::sync::{oneshot, OwnedSemaphorePermit, Semaphore};
+
+#[derive(Debug)]
+pub struct WalObjectStore {
+    object_store: Arc<dyn ObjectStore>,
+    file_notifier: Arc<dyn WalFileNotifier>,
+    /// Buffered wal ops go in here along with the state to track when to snapshot
+    flush_buffer: Mutex<FlushBuffer>,
+}
+
+impl WalObjectStore {
+    /// Creates a new WAL. Note that you will have to call replay and then start the flusher
+    /// to fully initialize the WAL.
```

Comment:
I wanted to be able to test a new wal without having the flusher run in the background. So I wanted to control the calls to flush so that I could validate along the way. Not sure how I should organize this to test it. The integration test of its usage should be higher up in the `WriteBufferImpl` and I have #25206 logged to track that.

Thoughts? I'm happy to change up the way it's organized a bit to make it clearer to use/understand.

Reply:
I thought of a way to do the testing the way I want and have this stuff in the call to new to make it a bit cleaner. Will push up an additional commit in a bit.

Reply:
Ok, updated here: 9b6e72a68004a447455c43b5bcfd97b9e6d5d98d